components/browser/icons/src/main/java/mozilla/components/browser/icons/generator/DefaultIconGenerator.kt

```diff
@@ -5,85 +5,97 @@
     override fun generate(context: Context, request: IconRequest): Icon {
-        val size = request.size.value.dpToPx(context.resources.displayMetrics)
+        val size = context.resources.getDimension(request.size.dimen)
+        val sizePx = size.toInt()
 
-        val bitmap = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888)
+        val bitmap = Bitmap.createBitmap(sizePx, sizePx, ARGB_8888)
         val canvas = Canvas(bitmap)
 
-        val backgroundColor = pickColor(request.url)
+        val backgroundColor = pickColor(context.resources, request.url)
 
         val paint = Paint()
         paint.color = backgroundColor
 
-        canvas.drawRoundRect(
-            RectF(0f, 0f, size.toFloat(), size.toFloat()), cornerRadius, cornerRadius, paint)
-
-        paint.color = textColor
+        val sizeRect = RectF(0f, 0f, size, size)
+        val cornerRadius = context.resources.getDimension(cornerRadiusDimen)
+        canvas.drawRoundRect(sizeRect, cornerRadius, cornerRadius, paint)
 
         val character = getRepresentativeCharacter(request.url)
 
         // The text size is calculated dynamically based on the target icon size (1/8th). For an icon
         // size of 112dp we'd use a text size of 14dp (112 / 8).
         val textSize = TypedValue.applyDimension(
             TypedValue.COMPLEX_UNIT_DIP,
-            size.toFloat() / 8.0f,
+            size / 8.0f,
             context.resources.displayMetrics
         )
 
+        paint.color = ContextCompat.getColor(context, textColorRes)
         paint.textAlign = Paint.Align.CENTER
         paint.textSize = textSize
         paint.isAntiAlias = true
 
         canvas.drawText(
             character,
-            canvas.width / 2.0f,
-            (canvas.height / 2.0f) - ((paint.descent() + paint.ascent()) / 2.0f),
+            canvas.width / 2f,
+            (canvas.height / 2f) - ((paint.descent() + paint.ascent()) / 2f),
             paint
         )
 
-        return Icon(bitmap = bitmap, color = backgroundColor, source = Icon.Source.GENERATOR)
+        return Icon(
+            bitmap = bitmap,
+            color = backgroundColor,
+            source = Icon.Source.GENERATOR
+        )
     }
 
     /**
      * Return a color for this [url]. Colors will be based on the host. URLs with the same host will
      * return the same color.
      */
-    internal fun pickColor(url: String): Int {
-        if (url.isEmpty()) {
-            return backgroundColors[0]
+    @ColorInt
+    internal fun pickColor(resources: Resources, url: String): Int {
+        return resources.obtainTypedArray(backgroundColorsRes).use { backgroundColors ->
+            if (url.isEmpty()) {
+                backgroundColors.getColor(0, 0)
+            } else {
+                val snippet = getRepresentativeSnippet(url)
+                val index = Math.abs(snippet.hashCode() % backgroundColors.length())
+
+                backgroundColors.getColor(index, 0)
+            }
```

Comment:
lint is complaining here:

> Error: This TypedArray should be recycled after use with #recycle()

may be a false positive since I assume that is `use` from androidx doing?

Reply:
It's a false positive. Somehow this is fine in other places but lint thinks its a problem here.